### DISCUSSION TITLE
How to layout timeseries so it can be optimally compressed with lz4

### DISCUSSION THREAD

**Discussion by MartyMcFlyInTheSky** [lz4/lz4#1354] [Q&A] [answered]
We use timeseries data and are thinking about switch to a new format. In that format, we're taking compression into account. How could we archieve maximum compression with lz4 with weather data, e.g. how can we layout the file so that streaks of zeroes and same values (e.g. -666) will compress optimally?

**Comment by t-mat** [comment, marked as answer]
Hi,

I'm not sure about your situation/constraint.  But since lz4 is a variant of LZ77, you can use basically same technique to improve the compression ratio.  It's able to compress consecutive same bytes pattern such as `uint8_t[] { 0, 0, 0, 0, ... }` or `uint8_t[] { 6, 0, 6, 0, ... }`. 

Also, you may be able to try the following topics:

1. (Basic) [`lz4` CLI](https://github.com/lz4/lz4/blob/5ff839680134437dbf4678f3d0c7b371d84f4964/programs/lz4.1.md#operation-modifiers).  You can try `-#` (compression level), `-B` (block size), `-BD` (block dependency), etc.  Also for small data, `-D` (dictionary) may work.

2. (Third party project) [c-blosc](https://github.com/Blosc/c-blosc) and [c-blosc2](https://github.com/Blosc/c-blosc2).  Blosc has many filters which (may) improve compression ratio.  Even if you don't use it directly, its filter methods will be a good hint.  For example, compress binary `float` or `double` more efficiently.

3. (Advanced) [lz4/examples](https://github.com/lz4/lz4/tree/5ff839680134437dbf4678f3d0c7b371d84f4964/examples).  These examples may help you to think about basic compression and data layout strategies.

**Reply by MartyMcFlyInTheSky** [reply]
Thank you so much, that helps a lot already :+1:
